### PR TITLE
Cache MessageBody across frames in same connection

### PR DESCRIPTION
**Pull request by benaadams**
Saves 1M object allocations and 48MB per million requests

### CODE REVIEW

**Comment by benaadams** [comment]
Would need more tests; but I doubt it will be merged any time soon...

File: src/Microsoft.AspNetCore.Server.Kestrel/Http/MessageBody.cs [hunk 1]
```diff
@@ -129,15 +134,32 @@ public Task Consume(CancellationToken cancellationToken = default(CancellationTo
...
                     return new ForContentLength(keepAlive, contentLength, context);
                 }
             }
 
             if (keepAlive)
             {
+                var messageBody = lastMessageBody as ForContentLength;
+                if (messageBody != null)
+                {
+                    messageBody.Reset(keepAlive, 0);
```
Comment (khellang) [inline src/Microsoft.AspNetCore.Server.Kestrel/Http/MessageBody.cs]: I don't get it... What's the point of resetting this when you're returning a new instance below anyway?

Comment (benaadams) [inline src/Microsoft.AspNetCore.Server.Kestrel/Http/MessageBody.cs]: Because I'm an idiot... :disappointed:

Comment (benaadams) [inline src/Microsoft.AspNetCore.Server.Kestrel/Http/MessageBody.cs]: Fixed idiocy... :smile:

File: src/Microsoft.AspNetCore.Server.Kestrel/Http/MessageBody.cs [hunk 2]
```diff
@@ -129,15 +135,34 @@ public Task Consume(CancellationToken cancellationToken = default(CancellationTo
...
+                    return messageBody;
+                }
+
                 return new ForContentLength(true, 0, context);
             }
 
+            if (lastMessageBody != null && lastMessageBody is ForRemainingData)
```
Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel/Http/MessageBody.cs]: `lastMessageBody is ForRemainingData`? isn't that impossible?

**Comment by halter73** [comment]
I don't think we're ready to take this kind of change for 1.0.0.

File: src/Microsoft.AspNetCore.Server.Kestrel/Http/MessageBody.cs (hunk 2 above)
Comment (benaadams) [inline src/Microsoft.AspNetCore.Server.Kestrel/Http/MessageBody.cs]: Probably

**Comment by benaadams** [comment]
> I don't think we're ready to take this kind of change for 1.0.0.

I assume not as it is a behavioural change

**Comment by halter73** [comment]
@benaadams Why is this a behavioral change? We're already caching `RequestBody`.

**Comment by benaadams** [comment]
Its reusing an object that's wasn't reused before; its not a great leap but it is a change

**Comment by halter73** [comment]
I'm closing this then. I don't think the impact is that significant. Post 1.0 we might revisit.

**Pull request closed without merging**